core/services/relay/evm/mercury/v2/data_source.go

```diff
@@ -0,0 +1,192 @@
+
+var maxInt192 = new(big.Int).Exp(big.NewInt(2), big.NewInt(191), nil)
+
+func NewDataSource(pr pipeline.Runner, jb job.Job, spec pipeline.Spec, feedID types.FeedID, lggr logger.Logger, rr chan pipeline.Run, enhancedTelemChan chan ocrcommon.EnhancedTelemetryMercuryData, fetcher LatestReportFetcher, linkFeedID, nativeFeedID types.FeedID) *datasource {
+	return &datasource{pr, jb, spec, feedID, lggr, rr, fetcher, linkFeedID, nativeFeedID, sync.RWMutex{}, enhancedTelemChan}
+}
+
+func (ds *datasource) Observe(ctx context.Context, repts ocrtypes.ReportTimestamp, fetchMaxFinalizedTimestamp bool) (obs relaymercuryv2.Observation, merr error) {
+	run, trrs, err := ds.executeRun(ctx)
+	if err != nil {
+		merr = fmt.Errorf("Observe failed while executing run: %w", err)
+		obs.BenchmarkPrice.Err = err
+		return
+	}
+	select {
+	case ds.runResults <- run:
+	default:
+		ds.lggr.Warnf("unable to enqueue run save for job ID %d, buffer full", ds.spec.JobID)
+	}
+
+	var parsed parseOutput
+	parsed, err = ds.parse(trrs)
+	if err != nil {
+		// This is not expected under normal circumstances
+		ds.lggr.Errorw("Observe failed while parsing run results", "err", err)
+		merr = fmt.Errorf("Observe failed while parsing run results: %w", err)
+		return
+	}
+	obs.BenchmarkPrice = parsed.benchmarkPrice
+
+	var wg sync.WaitGroup
+
+	if fetchMaxFinalizedTimestamp {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			obs.MaxFinalizedTimestamp.Val, obs.MaxFinalizedTimestamp.Err = ds.fetcher.LatestTimestamp(ctx)
+		}()
+	}
+
+	if ds.feedID == ds.linkFeedID {
+		// This IS the LINK feed, use our observed price
+		obs.LinkPrice.Val, obs.LinkPrice.Err = obs.BenchmarkPrice.Val, obs.BenchmarkPrice.Err
+	} else {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			obs.LinkPrice.Val, obs.LinkPrice.Err = ds.fetcher.LatestPrice(ctx, ds.linkFeedID)
+			if obs.LinkPrice.Val == nil && obs.LinkPrice.Err == nil {
+				ds.lggr.Warnw("Mercury server was missing LINK feed, falling back to max int192", "linkFeedID", ds.linkFeedID)
+				obs.LinkPrice.Val = maxInt192
+			}
+		}()
+	}
+
+	if ds.feedID == ds.nativeFeedID {
+		// This IS the native feed, use our observed price
+		obs.NativePrice.Val, obs.NativePrice.Err = obs.BenchmarkPrice.Val, obs.BenchmarkPrice.Err
+	} else {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			obs.NativePrice.Val, obs.NativePrice.Err = ds.fetcher.LatestPrice(ctx, ds.nativeFeedID)
+			if obs.NativePrice.Val == nil && obs.NativePrice.Err == nil {
+				ds.lggr.Warnw("Mercury server was missing native feed, falling back to max int192", "nativeFeedID", ds.nativeFeedID)
+				obs.NativePrice.Val = maxInt192
+			}
+		}()
+	}
+
+	wg.Wait()
+
+	// todo: implement telemetry
+	// if ocrcommon.ShouldCollectEnhancedTelemetryMercury(&ds.jb) {
+	// 	ocrcommon.EnqueueEnhancedTelem(ds.chEnhancedTelem, ocrcommon.EnhancedTelemetryMercuryData{
+	// 		TaskRunResults: trrs,
+	// 		Observation:    obs,
+	// 		RepTimestamp:   repts,
+	// 	})
+	// }
```

Comment:
Link ticket

Reply:
https://smartcontract-it.atlassian.net/browse/MERC-1388